Add process metadata for networking gadgets

alban opened the pull request:
# Add process metadata for networking gadgets

Supersedes #1193. This new version has support for both TCP and UDP.

This PR includes:
* a new bpf module "SocketEnricher" that uses kprobes to trace kernel functions related to bind, tcp connect, udp sendmsg and inet release to maintain a "sockets" bpf hash map containing the process metadata for each socket local endpoint.
* a bpf extension that provides helper functions such as gadget_skb_get_pid_tgid()
* updates to the dns and sni bpf modules to make use of the aforementioned bpf extension.

## How to use

```
$ sudo local-gadget trace dns -r docker
CONTAINER  PID      TID      COMM      QR NAMESERVER  TYPE      QTYPE  NAME
test       1291825  1291825  nslookup  Q  8.8.4.4     OUTGOING  A      wikipedia.org.
test       1291825  1291825  nslookup  R  8.8.4.4     HOST      A      wikipedia.org.
```

```
$ sudo local-gadget trace sni -r docker
CONTAINER   PID        TID        COMM     NAME
test        70022      70022      curl     wikipedia.org
```

```
$ kubectl gadget trace dns -n demo
NODE         NAMESPACE    POD          PID         TID         COMM        QR NAMESERVER      TYPE      QTYPE      NAME
minikube     demo         mypod        1285309     1285310     isc-net-00… Q  8.8.4.4         OUTGOING  A          inspektor-gadget.i…
minikube     demo         mypod        1285309     1285310     isc-net-00… R  8.8.4.4         HOST      A          inspektor-gadget.i…
```

```
$ kubectl gadget trace sni
NODE         NAMESPACE      POD        PID        TID        COMM      NAME
minikube     default        ubuntu     1123615    1123615    wget      wikimedia.org
```

## Testing done

See above.

## Limitations

- The sockets map can monitor up to max-entries=10240 sockets.
- Sockets created before the tracer are not detected.

burak-ok commented on pkg/gadgets/internal/socketenricher/tracer.go:
```diff
@@ -0,0 +1,215 @@
+
+	l, err = link.Kprobe("udpv6_sendmsg", se.objs.IgUdp6Sendmsg, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error opening udpv6_sendmsg ipv4 kprobe: %w", err)
```
copy/paste. Here: `ipv6`

burak-ok commented on integration/inspektor-gadget/integration_test.go:
```diff
@@ -547,6 +550,7 @@ func TestDns(t *testing.T) {
 				{
 					Event:      BuildBaseEvent(ns),
+					Comm:       "nslookup",
 					ID:         "0000",
```
Not because your changes: Shouldn't we leave this Field Initializer completly empty and set it in the `normalize` function to `""`?

alban replied:
Ok. That's how it was done in `integration/local-gadget/trace_dns_test.go` so we can do the same here.

burak-ok reviewed (approved):
Looks nice. Especially how little modifications is needed when it is used in a gadget.

mauriciovasquezbernal reviewed (commented):
First of all I want to say that this work is great! It's amazing how you were able to combine information from k[ret]probes to enrich network data. 

I added a bunch of comments regarding nits but I'd like to review the general approach first. I'm wondering if using [BPF Extensions](https://lore.kernel.org/bpf/9b204165-1d32-0e0d-ce19-10acaa45f9ed@fb.com/T/) is the better approach here. I understand this mechanism is great as it allows to change the behavior of an eBPF program at runtime, but I think it's an overkill for this case as we don't really need to change the behavior at runtime. I think an approach similar to what we use for filtering by mount namespace will work fine in this case:
1. Inject the socket-enricher programs: Those attach to the networking kprobes and keep the sockets map updated. 
2. Create a header file with the definition of such map and a helper function to loookup the socket. (the same `socket_lookup()` function you proposed here)
3. Replace the socket map of all loaded programs by the one in the socket-enricher 
4. Each gadget will use that helper function to get a socket_value and with that enrich the event

I tried it out in https://github.com/inspektor-gadget/inspektor-gadget/commit/2cf287ac9f1e36d249774f5c238e42555fce05cb and seems to work fine.

Advantages:
1. We don't need extension programs nor all its management around. (loading and attaching them). 
2. The performance is better as there is a single lookup to the sockets map 
3. We don't introduce a new kernel dependency on bpf extensions

What do you think? 

On a different topic. Now that we have the mount ns id, can we use it to enrich the events with container information from userspace? We'll be able to tell which container the event comes from. But I'm wondering what would happen to https://github.com/inspektor-gadget/inspektor-gadget/pull/1182, it seems it's not needed after this PR. :thinking: @blanquicet what do you think?

blanquicet commented on pkg/gadgets/trace/sni/tracer/tracer.go:
```diff
@@ -68,7 +67,12 @@ func parseSNIEvent(sample []byte) (*types.Event, error) {
-		Name: name,
+		Pid:       bpfEvent.Pid,
+		Tid:       bpfEvent.Tid,
+		MountNsID: bpfEvent.MountNsId,
```
Having the mountns, we could use the usual enricher and simplify the `ConnectToContainerCollection`. I did some experiments: https://github.com/inspektor-gadget/inspektor-gadget/commit/21fbac142060089d7c5a159d1445a9a3e45a3660. And it seems to work:

```bash
$ ./kubectl-gadget trace sni -A
NODE                  NAMESPACE             POD                   CONTAINER             PID         TID        COMM       NAME
master                                                                                  690         180537     containerd registry-1.docker.io
master                                                                                  690         181966     containerd auth.docker.io
master                                                                                  0           0                     www.inspektor-gadge…
master                                                                                  0           0                     www.inspektor-gadge…
master                default               mypod                 mypod                 182123      182123     wget       www.inspektor-gadge…
```

BTW, I noticed two things:

1. Name column needs more "weight" to avoid being truncated.
2. Some events were not enriched with PID info.

alban replied:
That's cool. Should I do that in the same PR?

> Name column needs more "weight" to avoid being truncated.

They already have `width:30` which should be enough for `www.inspektor-gadget.io`. I am not sure which value to choose.
I filed this issue for an alternative way: https://github.com/inspektor-gadget/inspektor-gadget/issues/1291

> Some events were not enriched with PID info. I suppose they were created before the tracer.

Correct. Because of this, we should not only rely on `MountNsID` but continue to have the current mechanism for namespace & pod.

blanquicet reviewed (commented):
🤯! Awesome Alban! I think it supersedes https://github.com/inspektor-gadget/inspektor-gadget/pull/1182 as it would add the container information (I added some comments for that) and even more information.

alban commented:
I am not sure it can fully replace #1182: I think adding the process metadata is best-effort and would not enrich events in every possible cases because of the following limitations:
* Sockets created before the tracer are not detected.
* The sockets map can monitor up to max-entries=10240 sockets.
* Packets sent without a tcp/udp socket (e.g. with a raw socket) would not be attached to a process
* Packets received without a tcp/udp socket (e.g. tcp-syn to a non-existent service, giving "connection refused") would not be attached to a process
* Packets received before applying a Netfilter mangle table or a port redirection might not match the port of the socket (not the typical case: CNI or kube-proxy would normally set iptables rules in the host network namespace).

alban commented on pkg/gadgets/internal/socketenricher/bpf/sockets-map.h:
```diff
@@ -0,0 +1,52 @@
+typedef __u32 ipv4_addr;
+
+struct sockets_key {
+	__u32 netns;
```
I initially used `__u64` but it does not work:
```
BPF_CORE_READ_INTO(&socket_key.netns, sock, __sk_common.skc_net.net, ns.inum);
```

`ns.inum` is an unsigned int (see [ns_common.h#L12](https://github.com/torvalds/linux/blob/f9ff5644bcc04221bae56f922122f2b7f5d24d62/include/linux/ns_common.h#L12) aka __u32). If we use `__u64`,  `BPF_CORE_READ_INTO` will copy `sizeof(*(dst))` bytes (8 bytes) so it will copy the `refcount_t count` fr the `struct ns_common` too.

mauriciovasquezbernal replied:
Perhaps we should switch to u32 for mntns too? It seems a bit weird to use different types for the same data in different places of the code. For sure, to be done in another PR.

blanquicet replied:
https://github.com/inspektor-gadget/inspektor-gadget/issues/1215

alban commented on pkg/gadgets/internal/socketenricher/bpf/sockets-map.bpf.c:
```diff
@@ -0,0 +1,268 @@
+
+	BPF_CORE_READ_INTO(&socket_key.netns, sock, __sk_common.skc_net.net, ns.inum);
+
+	socket_key.proto = BPF_CORE_READ_BITFIELD_PROBED(sock, sk_protocol);
```
> Why BPF_CORE_READ_BITFIELD_PROBED? 

It is done like this in pkg/gadgets/trace/bind/tracer/bpf/bindsnoop.bpf.c:
```
event.proto = BPF_CORE_READ_BITFIELD_PROBED(sock, sk_protocol);
```

> Isn't it an integer u16 sk_protocol?

True.
I will change it to BPF_CORE_READ.

alban commented on pkg/gadgets/internal/socketenricher/bpf/sockets-map.h:
```diff
@@ -0,0 +1,52 @@
+#define PACKET_HOST		0
+#endif
+
+#ifdef PACKET_OUTGOING
```
Same mistake in pkg/gadgets/trace/network/tracer/bpf/graph.c. I filed https://github.com/inspektor-gadget/inspektor-gadget/pull/1210 for it.

alban commented on pkg/gadgets/internal/socketenricher/bpf/sockets-map.h:
```diff
@@ -0,0 +1,52 @@
+	char task[TASK_COMM_LEN];
+
+	// 0 = client (connect)
+	// 1 = server (bind)
```
I am thinking to delete this field: it is currently not exposed to the user, and it is not well-defined for UDP.

alban commented on .gitignore:
```diff
@@ -18,11 +18,15 @@
 
 # Build artifacts
 *.o
 # keep all bpf objects generated with bpf2go
-!pkg/gadgets/*/*/tracer/*.o
+!pkg/gadgets/**/*bpfel.o
+!pkg/gadgets/**/*bpfel_x86.o
+!pkg/gadgets/**/*bpfel_arm64.o
```
Yes, it works. I'll take your suggestion.

mauriciovasquezbernal commented:
> I am not sure it can fully replace https://github.com/inspektor-gadget/inspektor-gadget/pull/1182: I think adding the process metadata is best-effort and would not enrich events in every possible cases because of the following limitations:

I agree. What about using both approaches to enrich then? We can first try with the mntns, if this is not available then we can use the netns?

alban commented:
> I'm wondering if using [BPF Extensions](https://lore.kernel.org/bpf/9b204165-1d32-0e0d-ce19-10acaa45f9ed@fb.com/T/) is the better approach here. I understand this mechanism is great as it allows to change the behavior of an eBPF program at runtime, but I think it's an overkill for this case as we don't really need to change the behavior at runtime. I think an approach similar to what we use for filtering by mount namespace will work fine in this case:
> [...]
> I tried it out in https://github.com/inspektor-gadget/inspektor-gadget/commit/2cf287ac9f1e36d249774f5c238e42555fce05cb and seems to work fine.

Fair enough. I actually implemented that way before and then switched to the bpf extension method because I found the API cleaner in the case of the byob gadget. But for the dns/sni gadgets, we don't need to have a stable API since they are internal gadgets. I will do it your way and remove the bpf extensions from this PR. We can revisit the bpf extension idea later for the byob gadget.

alban commented on pkg/gadgets/internal/socketenricher/bpf/extension.bpf.c:
```diff
@@ -0,0 +1,91 @@
+unsigned long long load_half(void *skb,
+			     unsigned long long off) asm("llvm.bpf.load.half");
+unsigned long long load_word(void *skb,
+			     unsigned long long off) asm("llvm.bpf.load.word");
```
Building on @mauriciovasquezbernal 's  2cf287a, they are now defined in only one place. It might not be the best location though. Could this be done in another PR?

mauriciovasquezbernal commented on pkg/gadgets/internal/socketenricher/bpf/sockets-map.bpf.c:
```diff
@@ -0,0 +1,231 @@
+	struct inet_sock *inet_sock = (struct inet_sock *)sock;
+	BPF_CORE_READ_INTO(&socket_key.netns, sock, __sk_common.skc_net.net, ns.inum);
+	socket_key.proto = BPF_CORE_READ(sock, sk_protocol);
+	socket_key.port = bpf_ntohs(BPF_CORE_READ(inet_sock, inet_sport));
```
I'm (for the nth time in the professional career) confused by endianess. In which endianess are ports saved on the sockets map? (host or network?)

If I use bpftool, I suppose it's saved in host as I can see `"port": 8000` after using `dig -b 172.17.0.3#8000 google.com`, but then, how does it work if there is not any endianess conversion in `gadget_socket_lookup()`?

alban replied:
In the kernel, `struct inet_sock` and `struct sock_common` have the following fields:
```
__be16 inet_sport;
__be16 skc_dport;
```
`__be16` => network endiannnes. Then we convert to host with bpf_ntohs, so the map stores values in host.

It is confusing because the kernel uses both host and network endianness in different fields...

alban replied:
I think I should add a comment near the map definition and where it is assigned.

mauriciovasquezbernal replied:
> The eBPF instructions load_half and load_word used in gadget_socket_lookup() automatically convert from network to host endianness 

:open_mouth: I didn't know it, now everything makes sense. 

I wondering if this makes sense to save them in networking order to avoid the endianess conversion for each packet, but I don't know how costly it is. Anyway, a comment could help a lot :D.

mauriciovasquezbernal reviewed (commented):
Some comments about the socket key not being unique. Those are not a blocker but I'd like to understand if there is a path forward to improve it in the future.

alban commented on pkg/gadgets/internal/socketenricher/bpf/sockets-map.bpf.c:
```diff
@@ -0,0 +1,231 @@
+	socket_value.pid_tgid = bpf_get_current_pid_tgid();
+	bpf_get_current_comm(&socket_value.task, sizeof(socket_value.task));
+
+	bpf_map_update_elem(&sockets, &socket_key, &socket_value, BPF_ANY);
```
Do you mean you prefer to give no information than potentially wrong information if the key is not unique?

mauriciovasquezbernal replied:
Exactly.

alban replied:
I am not sure which behaviour is more helpful. Do you know of applications using SO_REUSEPORT on Kubernetes? Since the netns is in the key, at least the 2 processes reusing the port would be in the same netns. But it could be different pods both using host networking. I'd like to have an example of such case to see if it's more practical for the user to have an idea of the application, or if it's better to not give any indications.

mauriciovasquezbernal replied:
What about checking if `SO_REUSEPORT` is enabled?, if it's then we can replace the element, otherwise we shouldn't replace the element because for some reason there was a clash. 
I also think this is not that important if we include family and source address in the key since I don't see other possibilities for collisions to happen other than `SO_REUSEPORT`.

alban commented on pkg/gadgets/internal/socketenricher/bpf/sockets-map.h:
```diff
@@ -0,0 +1,85 @@
+
+typedef __u32 ipv4_addr;
+
+struct sockets_key {
+	__u32 netns;
+
+	// proto is IPPROTO_TCP(6) or IPPROTO_UDP(17)
+	__u16 proto;
+	__u16 port;
+};
```
Thanks for the research!

We could probably add the family and the ip in the key to avoid clashes.

We have to be careful with IPv6 Addresses with Embedded IPv4 Addresses ([rfc4291 section-2.5.5](https://www.rfc-editor.org/rfc/rfc4291#section-2.5.5)). Even if we don't care about IPv6 and the process uses IPv4, the kernel might internally convert it to IPv6. So we have to handle IPv6 addresses anyway. We can use the idea from [is_ipv4_mapped_ipv6()](https://github.com/weaveworks/tcptracer-bpf/blob/d329ae7fb6fa90e4389075347f0754de2127336f/tcptracer-bpf.c#L132-L133) to check that.

About SO_REUSEPORT, I suggest to overwrite the map entry with the last info. For ingress, socket filter programs are too early in the network stack to know which socket the kernel is going to choose. I don't think it is a big deal if we give the wrong socket data. The two sockets are likely to be in the same process. But if one socket is released earlier than the other one, we would remove the entry from the map too early and then we would not get any process data for that connection. Do you think it is a problem in practice?

mauriciovasquezbernal replied:
I think it's fine to assume that all sockets sharing a port belong to the same application, that's what SO_REUSEPORT was designed for.

blanquicet commented:
> > I am not sure it can fully replace #1182: I think adding the process metadata is best-effort and would not enrich events in every possible cases because of the following limitations:
> 
> I agree. What about using both approaches to enrich then? We can first try with the mntns, if this is not available then we can use the netns.

It seems to be the compromise we were looking for. I just rebased https://github.com/inspektor-gadget/inspektor-gadget/pull/1182 to solve the merge conflicts, and it is ready again.

alban commented on pkg/gadgets/internal/socketenricher/bpf/sockets-map.h:
```diff
@@ -0,0 +1,85 @@
+{
+	struct sockets_key key = {0,};
+	key.netns = current_netns;
+	key.proto = load_byte(skb, ETH_HLEN + offsetof(struct iphdr, protocol));
```
I will add a check on IPv4.

For now both sni and dns check for IPv4 (`struct ethhdr->h_proto == ETH_P_IP`) and reject IPv6 (`struct ethhdr->h_proto == ETH_P_IPV6`).

alban commented on pkg/gadgets/internal/networktracer/tracer.go:
```diff
@@ -69,7 +74,25 @@ func newAttachment(
+		"current_netns": u32netns,
+	}
+
+	if err := spec.RewriteConstants(consts); err != nil && !strings.Contains(err.Error(), "spec is missing one or more constants") {
```
Done: https://github.com/cilium/ebpf/pull/902

alban replied:
Merged via https://github.com/cilium/ebpf/pull/904

alban replied:
I will first get the new cilium/ebpf via a separate PR: https://github.com/inspektor-gadget/inspektor-gadget/pull/1274

mauriciovasquezbernal replied:
I tried to update this code to use the `MissingConstantsError` type introduced in cilium/ebpf, however it's not that simple as it is defined as a structure, and then I'll have to use `errors.As()`. I was able to make it work in the end, but then I realized this approach doesn't work good as the map replacements logic will fail too:

```bash
$ sudo ./local-gadget trace dns
WARN[0000] Runtime enricher (cri-o): couldn't get current containers 
CONTAINER                                   QR NAMESERVER              TYPE      QTYPE      NAME                                        RCODE                 
err: container : start tracing container "local-registry": creating network tracer attachment for pid 3389: failed to create BPF collection: replacement map sockets not found in CollectionSpec
```

I think we have to change the strategy, first check if the given tracer is compiled with socket enricher support and then based on that replace the constants and the maps. I implemented this in https://github.com/inspektor-gadget/inspektor-gadget/commit/a9cb529acf8c020aa0cdcb1abea5ac8b1566a06e, I think it's fine to look for a map called `sockets`, if we think this could conflicts then we should switch the whole map name to something else.

blanquicet commented on pkg/gadgets/internal/networktracer/tracer.go:
```diff
@@ -69,7 +74,25 @@ func newAttachment(
 	}()
 
-	a.collection, err = ebpf.NewCollection(spec)
+	spec = spec.Copy()
```
A couple of eBPF questions 🙂 :

1) Why is this copy necessary?
2) Why we don't have to set the `current_netns` constant to the `SocketEnricher` specs, and it is ok to use the ones from dns and sni?

alban replied:
> Why is this copy necessary?

The `spec` pointer comes from the caller of NewTracer() and we don't own it, so we should not modify it.

Also, if 2 containers are created at the same time, 2 calls to `newAttachment()` could be done in parallel and if the same variable `spec` was modified in a different way for the 2 attachments, they might not get the correct value for current_netns...


> Why we don't have to set the current_netns constant to the SocketEnricher specs, and it is ok to use the ones from dns and sni?

`#include <sockets-map.h>` is done both in the sni/dns modules and in the SocketEnricher module so they both have the variable `current_netns`. However that variable is only used in `gadget_socket_lookup()` which is a static inline function only used in sni/dns. So `current_netns` is not used in the SocketEnricher module. Maybe I should move the variable declaration in the `#ifndef SOCKETS_MAP_IMPLEMENTATION` to make that clear. In any case, `current_netns` is not meaningful in the SocketEnricher module since there is only one instance of the SocketEnricher module, whereas there are `n` instances of the sni/dns modules (`n` being the number of containers' netns).

blanquicet replied:
Thanks for the explanation. It is all clear now!

> Maybe I should move the variable declaration in the #ifndef SOCKETS_MAP_IMPLEMENTATION to make that clear.
>
I think it is fine like it is.

blanquicet commented on pkg/gadgets/internal/socketenricher/tracer.go:
```diff
@@ -0,0 +1,150 @@
+
+	l, err = link.Kretprobe("inet_bind", se.objs.IgBindIpv4X, nil)
+	if err != nil {
+		return fmt.Errorf("error opening ipv4 kprobe: %w", err)
```
```suggestion
		return fmt.Errorf("error opening ipv4 kretprobe: %w", err)
```

blanquicet commented on pkg/gadgets/internal/socketenricher/tracer.go:
```diff
@@ -0,0 +1,150 @@
+
+	l, err = link.Kretprobe("inet6_bind", se.objs.IgBindIpv6X, nil)
+	if err != nil {
+		return fmt.Errorf("error opening ipv6 kprobe: %w", err)
```
```suggestion
		return fmt.Errorf("error opening ipv6 kretprobe: %w", err)
```

blanquicet commented on pkg/gadgets/internal/socketenricher/tracer_test.go:
```diff
@@ -0,0 +1,249 @@
+				}
+			}),
+		},
+		"tcp6": {
```
Why not UDP IPv6?

alban replied:
It did not work (the bind() syscall returns an error), I don't understand why. I got the tests from `pkg/gadgets/trace/bind/tracer/tracer_test.go` which also does not test UDP IPv6. Maybe the original author (@mauriciovasquezbernal ?) didn't write tests for UDP IPv6 because it also does not work there?

mauriciovasquezbernal replied:
It works on my machine :tm: 

```diff
diff --git pkg/gadgets/internal/socketenricher/tracer_test.go pkg/gadgets/internal/socketenricher/tracer_test.go
index 64cebd3f..e3d95a7d 100644
--- pkg/gadgets/internal/socketenricher/tracer_test.go
+++ pkg/gadgets/internal/socketenricher/tracer_test.go
@@ -145,6 +145,24 @@ func TestBindTracer(t *testing.T) {
                                }
                        }),
                },
+               "udp6": {
+                       generateEvent: bindSocketFn("::", unix.AF_INET6, unix.SOCK_DGRAM, 0),
+                       validateEvent: utilstest.ExpectAtLeastOneEvent(func(info *utilstest.RunnerInfo, port uint16) *socketEnricherMapEntry {
+                               return &socketEnricherMapEntry{
+                                       Key: socketenricherSocketsKey{
+                                               Netns:  uint32(netns),
+                                               Family: unix.AF_INET6,
+                                               Proto:  unix.IPPROTO_UDP,
+                                               Port:   port,
+                                       },
+                                       Value: socketenricherSocketsValue{
+                                               Mntns:   info.MountNsID,
+                                               PidTgid: uint64(uint32(info.Pid))<<32 + uint64(info.Tid),
+                                               Task:    stringToSlice("socketenricher."),
+                                       },
+                               }
+                       }),
+               },
        } {
                test := test
```

```bash
$ go test -exec sudo . -v -run 'TestBindTracer'
=== RUN   TestBindTracer
=== PAUSE TestBindTracer
=== CONT  TestBindTracer
=== RUN   TestBindTracer/tcp
=== PAUSE TestBindTracer/tcp
=== RUN   TestBindTracer/tcp6
=== PAUSE TestBindTracer/tcp6
=== RUN   TestBindTracer/udp6
=== PAUSE TestBindTracer/udp6
=== RUN   TestBindTracer/udp
=== PAUSE TestBindTracer/udp
=== CONT  TestBindTracer/tcp
=== CONT  TestBindTracer/udp6
=== CONT  TestBindTracer/udp
=== CONT  TestBindTracer/tcp6
--- PASS: TestBindTracer (0.00s)
    --- PASS: TestBindTracer/udp6 (2.92s)
    --- PASS: TestBindTracer/udp (3.01s)
    --- PASS: TestBindTracer/tcp6 (3.03s)
    --- PASS: TestBindTracer/tcp (3.22s)
PASS
ok  	github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher	3.242s
```

alban replied:
Thanks @mauriciovasquezbernal ! I will file a separate PR for a udp6 test in pkg/gadgets/trace/bind/tracer/tracer_test.go too.

alban replied:
Filed: https://github.com/inspektor-gadget/inspektor-gadget/pull/1290

blanquicet reviewed (commented):
Looking great! Some comments but mostly questions 🙂.

mauriciovasquezbernal reviewed (commented):
It's looking really good. Will do a final pass once these and other pending comments are handled. Thanks!

alban commented on docs/gadgets/trace/dns.md:
```diff
@@ -38,15 +38,17 @@ $ kubectl -n demo run mypod -it --image=wbitt/network-multitool -- /bin/sh
-minikube             demo                 mypod                R  8.8.4.4         HOST      AAAA        inspektor-gadget.i… NoError
-minikube             demo                 mypod                Q  8.8.4.4         OUTGOING  MX          inspektor-gadget.i…
-minikube             demo                 mypod                R  8.8.4.4         HOST      MX          inspektor-gadget.i… NoError
+NODE                 NAMESPACE            POD                  PID         TID         COMM        QR NAMESERVER      TYPE      QTYPE      NAME                RCODE
```
Updated!

alban commented:
The tests fail sometimes with:

```
=== CONT  TestTraceSni
    command.go:482: Stop command(StartTraceSniGadget)
    command.go:484: Command returned(StartTraceSniGadget):
        Error: failed to run gadget: timed out waiting for the condition

=== CONT  TestTraceDns
    command.go:482: Stop command(StartTraceDnsGadget)
    command.go:484: Command returned(StartTraceDnsGadget):
        Error: failed to run gadget: timed out waiting for the condition
```

It takes more than 3 seconds on my laptop to start the sni/dns tracers on an empty Minikube cluster (using bpftool prog to notice when the bpf socket filters are installed) and according to `/usr/share/bcc/tools/profile -p $PID_OF_GADGETTRACERMANAGER` the most frequent stack is in the socketenricher.

It seems there is a performance issue to understand before this PR can be merged.

<details>

```
    runtime.mapassign
    github.com/cilium/ebpf/btf.indexTypes
    github.com/cilium/ebpf/btf.(*Spec).Copy
    github.com/cilium/ebpf/btf.kernelSpec
    github.com/cilium/ebpf.applyRelocations
    local.github.com/cilium/ebpf.newProgramWithOptions
    github.com/cilium/ebpf.(*collectionLoader).loadProgram
    local.github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign.func1
    local.github.com/cilium/ebpf.assignValues
    github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.(*SocketEnricher).start
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.NewSocketEnricher
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/networktracer.NewTracer[go.shape.struct { github.com/inspektor-gadget/inspektor-gadget/pkg/types.Event; Pid uint32 "json:\"pid,omitempty\" column:\"pid,template:pid\""; Tid uint32 "json:\"tid,omitempty\" column:\"tid,template:pid\""; Comm string "json:\"comm,omitempty\" column:\"comm,template:comm\""; MountNsID uint64 "json:\"mountnsid,omitempty\" column:\"mntns,template:ns\""; ID string "json:\"id,omitempty\" column:\"id,width:4,fixed,hide\""; Qr github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/dns/types.DNSPktType "json:\"qr,omitempty\" column:\"qr,width:2,fixed\""; Nameserver string "json:\"nameserver,omitempty\" column:\"nameserver,template:ipaddr\""; PktType string "json:\"pktType,omitempty\" column:\"type,minWidth:7,maxWidth:9\""; QType string "json:\"qtype,omitempty\" column:\"qtype,minWidth:5,maxWidth:10\""; DNSName string "json:\"name,omitempty\" column:\"name,width:30\""; Rcode string "json:\"rcode,omitempty\" column:\"rcode,minWidth:8\""; Latency time.Duration "json:\"latency,omitempty\" column:\"latency,hide\"" }_0]
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/dns/tracer.NewTracer
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/dns.(*Trace).Start
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/dns.(*TraceFactory).Operations.func2
    github.com/inspektor-gadget/inspektor-gadget/pkg/controllers.(*TraceReconciler).Reconcile
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        4

    local.runtime.greyobject
    local.runtime.scanobject
    local.runtime.gcDrain
    local.runtime.gcBgMarkWorker.func2
    runtime.systemstack.abi0
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        4

    runtime.findObject
    local.runtime.scanobject
    local.runtime.gcDrain
    local.runtime.gcBgMarkWorker.func2
    runtime.systemstack.abi0
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        5

    runtime.findObject
    local.runtime.scanobject
    local.runtime.gcDrain
    local.runtime.gcBgMarkWorker.func2
    runtime.systemstack.abi0
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        5

    local.runtime.scanobject
    local.runtime.gcDrain
    local.runtime.gcBgMarkWorker.func2
    runtime.systemstack.abi0
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        5

    runtime.mapaccess1
    local.github.com/cilium/ebpf/btf.copier.copy
    github.com/cilium/ebpf/btf.copyTypes
    github.com/cilium/ebpf/btf.(*Spec).Copy
    github.com/cilium/ebpf/btf.kernelSpec
    github.com/cilium/ebpf.applyRelocations
    local.github.com/cilium/ebpf.newProgramWithOptions
    github.com/cilium/ebpf.(*collectionLoader).loadProgram
    local.github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign.func1
    local.github.com/cilium/ebpf.assignValues
    github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.(*SocketEnricher).start
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.NewSocketEnricher
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/networktracer.NewTracer[go.shape.struct { github.com/inspektor-gadget/inspektor-gadget/pkg/types.Event; Pid uint32 "json:\"pid,omitempty\" column:\"pid,template:pid\""; Tid uint32 "json:\"tid,omitempty\" column:\"tid,template:pid\""; Comm string "json:\"comm,omitempty\" column:\"comm,template:comm\""; MountNsID uint64 "json:\"mountnsid,omitempty\" column:\"mntns,template:ns\""; Name string "json:\"name,omitempty\" column:\"name,width:30\"" }_0]
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/sni/tracer.NewTracer
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/sni.(*Trace).Start
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/sni.(*TraceFactory).Operations.func2
    github.com/inspektor-gadget/inspektor-gadget/pkg/controllers.(*TraceReconciler).Reconcile
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        5

    runtime.mapaccess1
    local.github.com/cilium/ebpf/btf.copier.copy
    github.com/cilium/ebpf/btf.copyTypes
    github.com/cilium/ebpf/btf.(*Spec).Copy
    github.com/cilium/ebpf/btf.kernelSpec
    github.com/cilium/ebpf.applyRelocations
    local.github.com/cilium/ebpf.newProgramWithOptions
    github.com/cilium/ebpf.(*collectionLoader).loadProgram
    local.github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign.func1
    local.github.com/cilium/ebpf.assignValues
    github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.(*SocketEnricher).start
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.NewSocketEnricher
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/networktracer.NewTracer[go.shape.struct { github.com/inspektor-gadget/inspektor-gadget/pkg/types.Event; Pid uint32 "json:\"pid,omitempty\" column:\"pid,template:pid\""; Tid uint32 "json:\"tid,omitempty\" column:\"tid,template:pid\""; Comm string "json:\"comm,omitempty\" column:\"comm,template:comm\""; MountNsID uint64 "json:\"mountnsid,omitempty\" column:\"mntns,template:ns\""; ID string "json:\"id,omitempty\" column:\"id,width:4,fixed,hide\""; Qr github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/dns/types.DNSPktType "json:\"qr,omitempty\" column:\"qr,width:2,fixed\""; Nameserver string "json:\"nameserver,omitempty\" column:\"nameserver,template:ipaddr\""; PktType string "json:\"pktType,omitempty\" column:\"type,minWidth:7,maxWidth:9\""; QType string "json:\"qtype,omitempty\" column:\"qtype,minWidth:5,maxWidth:10\""; DNSName string "json:\"name,omitempty\" column:\"name,width:30\""; Rcode string "json:\"rcode,omitempty\" column:\"rcode,minWidth:8\""; Latency time.Duration "json:\"latency,omitempty\" column:\"latency,hide\"" }_0]
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/dns/tracer.NewTracer
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/dns.(*Trace).Start
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/dns.(*TraceFactory).Operations.func2
    github.com/inspektor-gadget/inspektor-gadget/pkg/controllers.(*TraceReconciler).Reconcile
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        6

    local.runtime.greyobject
    local.runtime.scanobject
    local.runtime.gcDrain
    local.runtime.gcBgMarkWorker.func2
    runtime.systemstack.abi0
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        6

    runtime.mapassign_faststr
    github.com/cilium/ebpf/btf.indexTypes
    github.com/cilium/ebpf/btf.(*Spec).Copy
    github.com/cilium/ebpf/btf.kernelSpec
    github.com/cilium/ebpf.applyRelocations
    local.github.com/cilium/ebpf.newProgramWithOptions
    github.com/cilium/ebpf.(*collectionLoader).loadProgram
    local.github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign.func1
    local.github.com/cilium/ebpf.assignValues
    github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.(*SocketEnricher).start
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.NewSocketEnricher
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/networktracer.NewTracer[go.shape.struct { github.com/inspektor-gadget/inspektor-gadget/pkg/types.Event; Pid uint32 "json:\"pid,omitempty\" column:\"pid,template:pid\""; Tid uint32 "json:\"tid,omitempty\" column:\"tid,template:pid\""; Comm string "json:\"comm,omitempty\" column:\"comm,template:comm\""; MountNsID uint64 "json:\"mountnsid,omitempty\" column:\"mntns,template:ns\""; Name string "json:\"name,omitempty\" column:\"name,width:30\"" }_0]
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/sni/tracer.NewTracer
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/sni.(*Trace).Start
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/sni.(*TraceFactory).Operations.func2
    github.com/inspektor-gadget/inspektor-gadget/pkg/controllers.(*TraceReconciler).Reconcile
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        7

    runtime.findObject
    local.runtime.scanobject
    local.runtime.gcDrain
    local.runtime.gcBgMarkWorker.func2
    runtime.systemstack.abi0
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        8

    local.runtime.scanobject
    local.runtime.gcDrain
    local.runtime.gcBgMarkWorker.func2
    runtime.systemstack.abi0
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        9

    local.runtime.scanobject
    local.runtime.gcDrain
    local.runtime.gcBgMarkWorker.func2
    runtime.systemstack.abi0
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        10

    runtime.mapassign
    github.com/cilium/ebpf/btf.indexTypes
    github.com/cilium/ebpf/btf.(*Spec).Copy
    github.com/cilium/ebpf/btf.kernelSpec
    github.com/cilium/ebpf.applyRelocations
    local.github.com/cilium/ebpf.newProgramWithOptions
    github.com/cilium/ebpf.(*collectionLoader).loadProgram
    local.github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign.func1
    local.github.com/cilium/ebpf.assignValues
    github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.(*SocketEnricher).start
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.NewSocketEnricher
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/networktracer.NewTracer[go.shape.struct { github.com/inspektor-gadget/inspektor-gadget/pkg/types.Event; Pid uint32 "json:\"pid,omitempty\" column:\"pid,template:pid\""; Tid uint32 "json:\"tid,omitempty\" column:\"tid,template:pid\""; Comm string "json:\"comm,omitempty\" column:\"comm,template:comm\""; MountNsID uint64 "json:\"mountnsid,omitempty\" column:\"mntns,template:ns\""; Name string "json:\"name,omitempty\" column:\"name,width:30\"" }_0]
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/sni/tracer.NewTracer
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/sni.(*Trace).Start
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/sni.(*TraceFactory).Operations.func2
    github.com/inspektor-gadget/inspektor-gadget/pkg/controllers.(*TraceReconciler).Reconcile
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        11

    runtime.mapaccess1
    local.github.com/cilium/ebpf/btf.copier.copy
    github.com/cilium/ebpf/btf.copyTypes
    github.com/cilium/ebpf/btf.(*Spec).Copy
    github.com/cilium/ebpf/btf.kernelSpec
    github.com/cilium/ebpf.applyRelocations
    local.github.com/cilium/ebpf.newProgramWithOptions
    github.com/cilium/ebpf.(*collectionLoader).loadProgram
    local.github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign.func1
    local.github.com/cilium/ebpf.assignValues
    github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.(*SocketEnricher).start
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/socketenricher.NewSocketEnricher
    github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/internal/networktracer.NewTracer[go.shape.struct { github.com/inspektor-gadget/inspektor-gadget/pkg/types.Event; Pid uint32 "json:\"pid,omitempty\" column:\"pid,template:pid\""; Tid uint32 "json:\"tid,omitempty\" column:\"tid,template:pid\""; Comm string "json:\"comm,omitempty\" column:\"comm,template:comm\""; MountNsID uint64 "json:\"mountnsid,omitempty\" column:\"mntns,template:ns\""; Name string "json:\"name,omitempty\" column:\"name,width:30\"" }_0]
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/sni/tracer.NewTracer
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/sni.(*Trace).Start
    local.github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-collection/gadgets/trace/sni.(*TraceFactory).Operations.func2
    github.com/inspektor-gadget/inspektor-gadget/pkg/controllers.(*TraceReconciler).Reconcile
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
    local.sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        12

    runtime.findObject
    local.runtime.scanobject
    local.runtime.gcDrain
    local.runtime.gcBgMarkWorker.func2
    runtime.systemstack.abi0
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        15

    local.runtime.scanobject
    local.runtime.gcDrain
    local.runtime.gcBgMarkWorker.func2
    runtime.systemstack.abi0
    local.runtime.goexit.abi0
    -                gadgettracerman (4158688)
        17
```

</details>

mauriciovasquezbernal commented:
Nice investigation @alban 

> Each call to btf.LoadKernelSpec() takes 237.193443ms.

It seems the library has a cache mechanism (https://github.com/cilium/ebpf/blob/dc0c82204f210e73e853bd0ad96060bb33ec3901/btf/btf.go#L290), isn't that working correctly?

We can move forward with this PR, but IMO it's something to be investigated in the library.

alban commented:
@mauriciovasquezbernal The cache is working correctly but it is the `Copy()` that is taking time:
https://github.com/cilium/ebpf/blob/dc0c82204f210e73e853bd0ad96060bb33ec3901/btf/btf.go#L318
```
return spec.Copy(), fallback, nil
```

alban commented on pkg/gadgets/internal/socketenricher/tracer.go:
```diff
@@ -0,0 +1,162 @@
+		return fmt.Errorf("failed to load asset: %w", err)
+	}
+
+	kernelSpec, err := btf.LoadKernelSpec()
+	if err != nil {
+		return fmt.Errorf("failed to load kernel spec: %w", err)
+	}
+	opts := &ebpf.CollectionOptions{
+		Programs: ebpf.ProgramOptions{
+			// Optimization: only one copy of kernelSpec necessary
+			// for all kprobes.
+			KernelTypes: kernelSpec,
+		},
```
With this optimization, the CI is finally green.

alban commented:
@mauriciovasquezbernal I think it is ready for another review.

TODO for future PRs:

- [ ] @blanquicet's suggestion to enrich with MountNsID ([comment](https://github.com/inspektor-gadget/inspektor-gadget/pull/1201#discussion_r1051245075) and https://github.com/inspektor-gadget/inspektor-gadget/commit/21fbac142060089d7c5a159d1445a9a3e45a3660)
- [ ] @mauriciovasquezbernal's [concern](https://github.com/inspektor-gadget/inspektor-gadget/pull/1201#discussion_r1053440134) about socket key uniqueness (same port but different family, SO_REUSEPORT, different addresses)

- [ ] @mauriciovasquezbernal's [suggestion](https://github.com/inspektor-gadget/inspektor-gadget/pull/1201#discussion_r1053497165) to avoid overwritting entries when it already exists (BPF_NOEXIST)

mauriciovasquezbernal commented on pkg/gadgets/internal/networktracer/tracer.go:
```diff
@@ -119,7 +143,13 @@ func NewTracer[Event any](
 ) *Tracer[Event] {
+	socketEnricher, err := socketenricher.NewSocketEnricher()
+	if err != nil {
+		// Non fatal: an older kernel might not support bpf extensions
```
After commenting on https://github.com/inspektor-gadget/inspektor-gadget/pull/1201/files#r1095999193, I checked this and I think we can fix both issues with the same change: Only create the socker enricher if this is use by the gadget, I tried some experiments in https://github.com/inspektor-gadget/inspektor-gadget/commit/5a74593a841da7ce5b8351e9b73cb67465d3bd2c.

alban replied:
The SocketEnricher uses CO-RE, so your patch 5a74593a841da7ce5b8351e9b73cb67465d3bd2c would break the dns and sni gadgets on older kernels without BTF/btfhub support. Is it ok?

mauriciovasquezbernal reviewed (commented):
It's looking even better now. I added some comments about how to handle the case when a tracer doesn't use this socket enricher, besides that LGTM!

alban commented:
@mauriciovasquezbernal I picked up your commit 5a74593a841da7ce5b8351e9b73cb67465d3bd2c but making the error (https://github.com/inspektor-gadget/inspektor-gadget/commit/5a74593a841da7ce5b8351e9b73cb67465d3bd2c#r99476727) non-fatal. Does it look ok?

mauriciovasquezbernal reviewed (approved):
LGTM! Great work @alban, it's really a nice new feature that makes Inspektor Gadget better. 

I added some nits, feel free to handle or reject them at your discretion.

blanquicet commented:
> TODO for future PRs:
> 
> * [ ]  @blanquicet's suggestion to enrich with MountNsID ([comment](https://github.com/inspektor-gadget/inspektor-gadget/pull/1201#discussion_r1051245075) and [21fbac1](https://github.com/inspektor-gadget/inspektor-gadget/commit/21fbac142060089d7c5a159d1445a9a3e45a3660))
>
This is being managed in https://github.com/inspektor-gadget/inspektor-gadget/pull/1281
